### ISSUE TITLE
Possible memory leak

### ISSUE THREAD

**Issue by tbflw** [zaggino/z-schema#192]
We use node-swagger in our node production services and since late August we have been hunting down a memory leak in all services recently deployed. It turns out that the leak is caused by a new release of z-schema (3.18.3) being dragged in as a downstream dependency of swagger-node-runner. We suspect that the implementation of ReferenceCache in [this commit](https://github.com/zaggino/z-schema/commit/b1ea164b7acba124de45d149b42279411e6eefeb) is the cause. 

Locking down the version of z-schema to 3.18.2 via shrinkwrap fixed the memory leak. We have reported the memory leak in swagger-node-runner with documentation of the leak from our production environment here: https://github.com/theganyo/swagger-node-runner/issues/110

**Comment by zaggino** [comment]
cc @jfromaniello

**Comment by zaggino** [comment]
@tbflw do you think using `WeakMap` instead of `Map` would solve the issue? Or would be better to just revert the commit?

**Comment by jfromaniello** [comment]
My change will leak if you do something like this:

```
validator.validate(object, { .. });
```

i.e. if you define the schema each time you validate.

This was a problem on previous version of this library but it was "solved" in recent versions by doing `_.deepEqual` https://www.npmjs.com/package/lodash.isequal . I know this because I was using the old version and I found the leak in some path of our application.

**deepEqual does prevent this type of leak but at the cost of precious CPU cycles by having to ite rate through all the schemas and each property of each schema.**

Definitely a WeakMap will solve this issue, but on the other hand we will be compiling and validating the schema every time.

I am trying to think other ways to solve this issue. I'd like this library to be fast and also to prevent the user for shooting their foot.

**Comment by tbflw** [comment]
I certainly agree on the requirement of the lib being fast and compiling and validating the schema upon every request does not seem like a good idea. On the other hand, the leak is bad. I haven't looked into how validation is called from upstream (sway, swagger-node-runer). Would it be possible for you guys to investigate this or collaborate with those projects on finding a solution?

**Comment by tbflw** [comment]
I had a look at how sway uses z-schema, and it is indeed being used in the way you suggested. In [sway/lib/helpers.js](https://github.com/apigee-127/sway/blob/6a5182bdf0beadd25dd90729ed16880ea5cac276/lib/helpers.js) lines 434 - 451:
```js
module.exports.validateAgainstSchema = function (validator, schema, value) {
  schema = _.cloneDeep(schema); // Clone the schema as z-schema alters the provided document

  var response = {
    errors: [],
    warnings: []
  };

  if (!validator.validate(value, schema)) {
    response.errors = _.map(validator.getLastErrors(), function (err) {
      normalizeError(err);

      return err;
    });
  }

  return response;
};
```
@jfromaniello Can you suggest a possible change in sway to better use z-schema avoiding validating the schema on each call and avoiding the leak at the same time?

**Comment by zaggino** [comment]
`npm deprecate z-schema@3.18.3 "memory leak bug, revert to 3.18.2"`

**Comment by osher** [comment]
+1

**Comment by WaldoJeffers** [comment]
What's the current status of this bug? Is this fixed in z-schema v.3.19.0? Or does something needs to be changed in sway? I couldn't find any issues related to this on the sway repo.

**Comment by zaggino** [comment]
Reverted the problematic commit and fixed since `3.18.4`

**Comment by WaldoJeffers** [comment]
Thanks @zaggino :)
